--- a/drivers/gpu/drm/sun4i/sun4i_tcon.h
+++ b/drivers/gpu/drm/sun4i/sun4i_tcon.h
@@ -53,6 +53,7 @@
 #define SUN4I_TCON0_CTL_TCON_ENABLE			BIT(31)
 #define SUN4I_TCON0_CTL_IF_MASK				GENMASK(25, 24)
 #define SUN4I_TCON0_CTL_IF_8080				(1 << 24)
+#define SUN4I_TCON0_CTL_SWAP				BIT(23)
 #define SUN4I_TCON0_CTL_CLK_DELAY_MASK			GENMASK(8, 4)
 #define SUN4I_TCON0_CTL_CLK_DELAY(delay)		((delay << 4) & SUN4I_TCON0_CTL_CLK_DELAY_MASK)
 #define SUN4I_TCON0_CTL_SRC_SEL_MASK			GENMASK(2, 0)
--- a/drivers/gpu/drm/sun4i/sun4i_tcon.c
+++ b/drivers/gpu/drm/sun4i/sun4i_tcon.c
@@ -38,6 +38,8 @@
 #include "sun8i_tcon_top.h"
 #include "sunxi_engine.h"
 
+static int swap_b_r = 0;
+
 static struct drm_connector *sun4i_tcon_get_connector(const struct drm_encoder *encoder)
 {
 	struct drm_connector *connector;
@@ -501,6 +503,15 @@
 			   SUN4I_TCON0_CTL_CLK_DELAY_MASK,
 			   SUN4I_TCON0_CTL_CLK_DELAY(clk_delay));
 
+	/* Swap Red & Blue of LCD */
+	if(swap_b_r){
+		printk("srgn: do swap.");
+		regmap_update_bits(tcon->regs, SUN4I_TCON0_CTL_REG,
+			   SUN4I_TCON0_CTL_SWAP,
+			   SUN4I_TCON0_CTL_SWAP);
+	}
+
+
 	/*
 	 * This is called a backporch in the register documentation,
 	 * but it really is the back porch + hsync
@@ -1329,6 +1340,13 @@
 			return ret;
 	}
 
+	// no more dirty things happen! oh yeah!
+	swap_b_r = of_property_read_bool(node, "srgn,swap-b-r");
+	
+	if(swap_b_r){
+		printk("srgn: swap r&b as per device tree.\n");
+	}
+	
 	return component_add(&pdev->dev, &sun4i_tcon_ops);
 }
 
